### PR TITLE
Final step of Week 3 PR! (Step 6 Part 2)

### PR DESCRIPTION
**Pull request by mmurray22**
Final step of Week 3 PR! (Step 6 Part 2)

### CODE REVIEW

**Comment by ccondit** [comment]
Bianca & Ihsan - please review and after you've LGTMed add Kaveh and I back for final approval.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    int numDeleteComments = 1;
+    Query query = new Query(COMMENT_TABLE_NAME ).addSort(TIMESTAMP_COLUMN_NAME, SortDirection.ASCENDING);
+    PreparedQuery results = datastore.prepare(query);
+    String inputNumDelete = request.getParameter("delete-num");
+    if (inputNumDelete != null && !inputNumDelete.isEmpty()) {
+        numDeleteComments = Integer.parseInt(inputNumDelete);
+    }
+    int counter = 0;
```
Comment (honeybeebaby) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java]: does ```int counter``` need to be here? i dont see it doing anything else in the ```doPost```, your ```for``` loop seems to looping through entities and not counting anything.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import java.io.IOException;
+import com.google.gson.Gson;
```
Comment (honeybeebaby) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java]: line 24 should be the first ```import```

**Review by honeybeebaby — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/index.html [hunk 3]
```diff
@@ -69,6 +69,10 @@ <h1>Michaela Murray's Portfolio</h1>
       <br/><br/>
       <input type="submit"/>
     </form>
+    <form method="POST" action="/delete-data">
+      <input type="number" name="delete-num" />
```
Comment (ihsan314) [inline portfolio/src/main/webapp/index.html:99]: Wouldn't be a bad idea to have a label showing the user this textbox contains the number of comments they would like to delete.

File: portfolio/src/main/webapp/script.js [hunk 4]
```diff
@@ -71,10 +71,6 @@ function contact() {
 	console.log('Contact!'); 
 }
 
-function updateNumber() {
-    getComments();
-}
-
 function getComments() {
 	fetch('/data?max-num=' + (document.getElementById("numComments").value)).then(response => response.json()).then((comments) => {
```
Comment (ihsan314) [inline portfolio/src/main/webapp/script.js]: Maybe consider restructuring this line like this:
```
fetch('/data?max-num=' + (document.getElementById("numComments").value))
  .then(response => response.json())
  .then((comments) => {
```
so that you don't have an overly long line.

**Review by ihsan314 — CHANGES_REQUESTED** [review]
Overall pretty good, found a couple things that might have better alternatives.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java (hunk 2 above)
Comment (honeybeebaby) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java]: wait im wrong it should be switched with line 23 to be in alphabetical order, sorry.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java (hunk 1 above)
Comment (mmurray22) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java]: yes indeed! thanks-updated!

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java (hunk 2 above)
Comment (mmurray22) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java]: thanks! updated!

File: portfolio/src/main/webapp/script.js (hunk 4 above)
Comment (mmurray22) [reply portfolio/src/main/webapp/script.js]: thanks! updated!

File: portfolio/src/main/webapp/index.html (hunk 3 above)
Comment (mmurray22) [reply portfolio/src/main/webapp/index.html:99]: thanks! updated!

**Review by ihsan314 — APPROVED** [review]
LGTM

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java (hunk 2 above)
Comment (ihsan314) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java]: Chris wrote a [bash script](https://gist.github.com/ccondit/55f6760a001a9aa4e2167e81ea9bd73b) tool to help with this in the future.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/** Servlet that returns some example content. */
+@WebServlet("/delete-data")
+public class DeleteComments extends HttpServlet {
+  DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+  private static final String COMMENT_TABLE_NAME = "Comment";
+  private static final String TIMESTAMP_COLUMN_NAME = "submit_time";
```
Comment (ccondit) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java]: Just something to think about - now these constants are spread across two classes. How will they evolve over time? Might be nice to refactor them to a common location...

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/** Servlet that returns some example content. */
+@WebServlet("/delete-data")
+public class DeleteComments extends HttpServlet {
+  DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
```
Comment (ccondit) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java]: nit - reduce visibility as much as possible:
```suggestion
  private final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
```

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  private static final String COMMENT_TABLE_NAME = "Comment";
+  private static final String TIMESTAMP_COLUMN_NAME = "submit_time";
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    int numDeleteComments = 1;
```
Comment (ccondit) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java]: nit - this might be clearer as `defaultNumCommentsToDelete`? Likewise with the `inputNumDelete` below.

also consider moving the defaultNumCommentsToDelete to a constant.

File: portfolio/src/main/webapp/index.html (hunk 3 above)
Comment (ccondit) [reply portfolio/src/main/webapp/index.html:99]: Can you associate this [label with the input](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label)? Either explicitly or implicitly. This will help accessibility.

File: portfolio/src/main/webapp/script.js [hunk 8]
```diff
@@ -72,12 +72,20 @@ function contact() {
...
-      for (comment of comments) {
-        bodyElement.innerHTML += comment + '<br>';
-      }
+	fetch('/data?max-num=' + (document.getElementById("numComments").value))
+        .then(response => response.json())
+        .then((comments) => {
+            const bodyElement = document.body;
+            console.log('Comments: ', comments);
+            bodyElement.innerHTML += 'Here are all the comments: <br>';
+            for (comment of comments) {
```
Comment (ccondit) [inline portfolio/src/main/webapp/script.js]:
```suggestion
            for (const comment of comments) {
```

**Review by ccondit — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java (hunk 5 above)
Comment (mmurray22) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java]: ah good point! will address!

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java (hunk 6 above)
Comment (mmurray22) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java]: noted! updated!

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java (hunk 7 above)
Comment (mmurray22) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java]: ah ok noted! if I move defaultNumCommentsToDelete to a constant, how can I change it later when I get the value?

File: portfolio/src/main/webapp/index.html (hunk 3 above)
Comment (mmurray22) [reply portfolio/src/main/webapp/index.html:99]: It looks like the label is not present in this particular thread, but it should be added :)

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java (hunk 7 above)
Comment (mmurray22) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java]: hmm it seems like we are talking about two different models. Currently, I'm using numDeleteComments as the baseline, and if the user updates the delete num value to a valid number, then I change that value to the user requested value. Do you think there is a better design here?

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 9]
```diff
@@ -34,9 +34,9 @@
 /** Servlet that returns some example content. */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-  private static final String COMMENT_TABLE_NAME = "Comment";
-  private static final String COMMENT_COLUMN_NAME = "text";
-  private static final String TIMESTAMP_COLUMN_NAME = "submit_time";
+  public static final String COMMENT_TABLE_NAME = "Comment";
+  public static final String COMMENT_COLUMN_NAME = "text";
+  public static final String TIMESTAMP_COLUMN_NAME = "submit_time";
```
Comment (ccondit) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: nit - Since these servlets are all in the same package, you can use [package private visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html). In general, prefer the least level of visibility possible.

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java [hunk 10]
```diff
@@ -33,21 +33,21 @@
...
-  private static final String TIMESTAMP_COLUMN_NAME = "submit_time";
+  private final DatastoreService dataStore = DatastoreServiceFactory.getDatastoreService();
+  private DataServlet dataServlet = new DataServlet();
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     int numDeleteComments = 1;
-    Query query = new Query(COMMENT_TABLE_NAME ).addSort(TIMESTAMP_COLUMN_NAME, SortDirection.ASCENDING);
-    PreparedQuery results = datastore.prepare(query);
+    Query query = new Query(dataServlet.COMMENT_TABLE_NAME )
```
Comment (ccondit) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java]: Since the constants are static members of `DataServlet` - you don't need to create an instance to get at them.
```suggestion
    Query query = new Query(DataServlet.COMMENT_TABLE_NAME )
```

Can you also remove the `DataServlet` instantiation on line 37?

File: portfolio/src/main/webapp/script.js [hunk 11]
```diff
@@ -66,13 +66,16 @@ function getComments() {
...
-    console.log("Starting to delete comments!");
-    fetch('/delete-data');
-    console.log("Complete!");
+/*
+ * Creates a map and adds it to the page. 
+ */
+function createMap() {
+  const map = new google.maps.Map(
+      document.getElementById('map'),
+      {center: {lat: 37.422, lng: -122.084}, zoom: 16});
```
Comment (ccondit) [inline portfolio/src/main/webapp/script.js]: Is this related to this PR?

**Review by ccondit — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/script.js (hunk 11 above)
Comment (mmurray22) [reply portfolio/src/main/webapp/script.js]: this is merged in from the master branch

Comment (mmurray22) [reply portfolio/src/main/webapp/script.js]: should I delete it?

Comment (mmurray22) [reply portfolio/src/main/webapp/script.js]: oh I'll try updating Week-6-1 to see if that helps!

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java (hunk 9 above)
Comment (mmurray22) [reply portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: oooh cool I did not know about that! updated!

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java (hunk 10 above)
Comment (mmurray22) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java]: oh whoops! thanks!

File: portfolio/src/main/webapp/script.js (hunk 8 above)
Comment (ccondit) [reply portfolio/src/main/webapp/script.js]: doesn't look like this got resolved...

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java (hunk 2 above)
Comment (ccondit) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java]: Can you try running this file through the bash script? There are now unused imports (which it will hopefully remove) and the formatting looks slightly off...

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java (hunk 7 above)
Comment (ccondit) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java]: This comment was more about the variable name. `numDeleteComments` reads more like "number of deleted comments" or "the number of deleted comments" (past tense). Feel free to be expressive with variable naming s.t. the name really reflects the purpose of the variable.

**Review by ccondit — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    int numCommentsToDelete = 1;
+    Query query = new Query(DataServlet.COMMENT_TABLE_NAME)
+                            .addSort(DataServlet.TIMESTAMP_COLUMN_NAME, SortDirection.ASCENDING);
+    PreparedQuery results = dataStore.prepare(query);
+    String inputNumDelete = request.getParameter("delete-num");
+    if (inputNumDelete != null && !inputNumDelete.isEmpty()) {
+        numCommentsToDelete = Integer.parseInt(inputNumDelete);
```
Comment (kaveh096) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java:41]: This method could throw `NumberFormatException`. So you need to handle it somehow. Either catch it and ignore it (delete nothing) or return proper error back to user.

**Review by kaveh096 — APPROVED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java (hunk 12 above)
Comment (mmurray22) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java:41]: alrighty! will updated! thanks!

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    int numCommentsToDelete = 1;
+    Query query = new Query(DataServlet.COMMENT_TABLE_NAME)
+                            .addSort(DataServlet.TIMESTAMP_COLUMN_NAME, SortDirection.ASCENDING);
+    PreparedQuery results = dataStore.prepare(query);
+    String inputNumDelete = request.getParameter("delete-num");
+    if (inputNumDelete != null && !inputNumDelete.isEmpty()) {
+        try{
```
Comment (ccondit) [inline portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java]: this formatting looks off - did you try running this class through the bash script?

File: portfolio/src/main/webapp/index.html (hunk 3 above)
Comment (ccondit) [reply portfolio/src/main/webapp/index.html:99]: did this get done? looks like the label still doesn't have a `for` attribute...

**Review by ccondit — CHANGES_REQUESTED** [review]

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java (hunk 13 above)
Comment (mmurray22) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java]: I did :/ commented in the gc about problem with the shell...

File: portfolio/src/main/webapp/index.html (hunk 3 above)
Comment (mmurray22) [reply portfolio/src/main/webapp/index.html:99]: should be addressed now!

File: portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java (hunk 13 above)
Comment (ccondit) [reply portfolio/src/main/java/com/google/sps/servlets/DeleteComments.java]: in the bash script s/1.7/1.8/g. should solve the problem. i think formatting is my last nit here.

**Review by ccondit — CHANGES_REQUESTED** [review]

**Review by ccondit — APPROVED** [review]

**Review by honeybeebaby — APPROVED** [review]
LGTM!
